src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1895 @@
+                                  "Job terminated abnormally with signal %s.",
+                                  signal_to_string(si->si_status));
+        } else if (si->si_status != EXIT_SUCCESS)
+                if (j->status_errno != 0)
+                        sd_bus_error_set_errno(&error, j->status_errno);
+                else
+                        sd_bus_error_setf(&error, SD_BUS_ERROR_FAILED,
+                                          "Job failed with exit code %i.", si->si_status);
+        else {
+                r = job_parse_child_output(TAKE_FD(j->stdout_fd), &json);
+                if (r < 0)
+                        sd_bus_error_set_errnof(&error, r, "Failed to parse JSON: %m");
+        }
+
+        r = sd_bus_emit_signal(
+                        j->manager->bus,
+                        "/org/freedesktop/sysupdate1",
+                        "org.freedesktop.sysupdate1.Manager",
+                        "JobRemoved",
+                        "uoi",
+                        j->id,
+                        j->object_path,
+                        j->status_errno != 0 ? -j->status_errno : si->si_status);
+        if (r < 0)
+                log_warning_errno(r, "Cannot emit JobRemoved message, ignoring: %m");
+
+        if (j->dbus_msg && j->complete_cb) {
+                if (sd_bus_error_is_set(&error)) {
+                        log_warning("Bus error occurred, ignoring callback for job: %s", error.message);
+                        sd_bus_reply_method_error(j->dbus_msg, &error);
+                } else {
+                        r = j->complete_cb(j->dbus_msg, j, json, &error);
+                        if (r < 0 || sd_bus_error_is_set(&error)) {
+                                log_warning_errno(r, "Error during execution of job callback: %m");
+                                sd_bus_reply_method_errno(j->dbus_msg, r, &error);
+                        }
+                }
+        }
+
+        job_free(j);
+        return 0;
+}
+
+static inline const char* sysupdate_binary_path(void) {
+        return secure_getenv("SYSTEMD_SYSUPDATE_PATH") ?: SYSTEMD_SYSUPDATE_PATH;
+}
+
+static int target_get_argument(Target *t, char **ret) {
+        _cleanup_free_ char *target_arg = NULL;
+
+        assert(t);
+        assert(ret);
+
+        if (t->class != TARGET_HOST) {
+                if (t->class == TARGET_COMPONENT)
+                        target_arg = strjoin("--component=", t->name);
+                else if (IN_SET(t->image_type, IMAGE_DIRECTORY, IMAGE_SUBVOLUME))
+                        target_arg = strjoin("--root=", t->path);
+                else if (IN_SET(t->image_type, IMAGE_RAW, IMAGE_BLOCK))
+                        target_arg = strjoin("--image=", t->path);
+                else
+                        assert_not_reached();
+                if (!target_arg)
+                        return -ENOMEM;
+        }
+
+        *ret = TAKE_PTR(target_arg);
+        return 0;
+}
+
+static int job_start(Job *j) {
```

Comment:
this is another one of those functions that sometimes logs and sometimes doesn't. please clean that up. i.e. we want that for each error we get exactly one log message, not zero and not two.